UIA objects: introduce a global suggestion list item class for Windows 10. re #6241

josephsl opened the pull request:
Windows 10 uses suggestions list for vairous things, including Start menu suggestions, Store recommendations, Settings app and others. Thus introduce NVDAObjects.UIA.SuggestionListItem, which derives its power from searchui.py version (searchui.py version is gone).

Suggested what's new entry:
- Category: not sure
  In Windows 10, you can use up and down arrow keys to review suggestion results (e.g. Settings app suggestions and Store recommendations).

Thanks.

josephsl committed:
UIA handler: add proper copyright header (based on Git log archive)

josephsl committed:
UIA objects: introduce a global suggestion list item class for Windows 10. re #6241

Windows 10 uses suggestions list for vairous things, including Start menu suggestions, Store recommendations, Settings app and others. Thus introduce NVDAObjects.UIA.SuggestionListItem, which derives its power from searchui.py version (searchui.py version is gone).

derekriemer commented:
Hi, maybe we should create a UIA  behaviors package?

Sent from a mobile device.
Please disregard errors as this is a smaller device.

> On Aug 16, 2016, at 21:25, Joseph Lee <notifications@github.com> wrote:
> 
> Windows 10 uses suggestions list for vairous things, including Start menu suggestions, Store recommendations, Settings app and others. Thus introduce NVDAObjects.UIA.SuggestionListItem, which derives its power from searchui.py version (searchui.py version is gone).
> 
> Suggested what's new entry:
> 
> Category: not sure In Windows 10, you can use up and down arrow keys to review suggestion results (e.g. Settings app suggestions and Store recommendations).
> Thanks.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/nvaccess/nvda/pull/6274
> 
> Commit Summary
> 
> UIA objects: introduce a global suggestion list item class for Windows 10. re #6241
> File Changes
> 
> M source/NVDAObjects/UIA/__init__.py (19)
> M source/appModules/searchui.py (18)
> Patch Links:
> 
> https://github.com/nvaccess/nvda/pull/6274.patch
> https://github.com/nvaccess/nvda/pull/6274.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

josephsl committed:
SearchUI: remove unneeded imports

josephsl commented:
Hi,

UIA behaviors: I'm thinking not.

josephsl committed:
UIA: add support for Controller For property. re #6241

Controller For property is used by an element to let another element depend on values given by the source element, useful if suggestions should be listed for a search field among other possibilities.
For now, Controller For property has been added to property to event names map in UIA Handler. A UIA object that handles this event (a search box) is next.

josephsl committed:
Revert "UIA: add support for Controller For property. re #6241"

This reverts commit a3b1ce57408fc571b1f92aa38901c243a8171fad.

josephsl committed:
UIA: add support for Controller For property. re #6241

Controller For property is used by an element to let another element depend on values given by the source element, useful if suggestions should be listed for a search field among other possibilities.
For now, Controller For property has been added to property to event names map in UIA Handler. A UIA object that handles this event (a search box) is next.

josephsl committed:
NVDAObjects.UIA: introduce Search Field that'll handle controller for and announce appearance of suggestions. re #6241

Coming from Windows 10 App Essentials add-on: a Search Field is now available that'll detect controller for property event and announce either 'suggestions' or 'suggestions closed' if suggestions appear or disappear, respectively. This takes care of both Start Menu and other fields in Windows 10 (works best in Anniversary Update and later).

josephsl commented:
@MichaelDCurran, I'd like to request a review please. Thanks.

michaelDCurran commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -369,6 +370,14 @@ def findOverlayClasses(self,clsList):
+		if self.UIAElement.cachedAutomationID in ("TextBox", "SearchTextBox"):
+			clsList.append(SearchField)
+		try:
+			if self.parent.UIAElement.cachedAutomationId.lower()=="suggestionslist":
```
This could be very costly and could end up being recursive -- I had made the same mistake with Microsoft Edge code earlier this year. Items within a suggestionList don't have a useful UIA className or automationID?

michaelDCurran replied:
If it is absolutely necessary to check the parent, then you can use raw UIA calls. Using the UIAHandler.handler.baseTreeWalker and get the parent element for self.UIAElement and check its automationID.

josephsl replied:
I see. I'll read EdgeRS1 branch to see what I can learn from there before I proceed. Thanks.

michaelDCurran commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -998,3 +1007,32 @@ def event_nameChange(self):
+	def event_UIA_controllerFor(self):
+		# Only useful if suggestions appear and disappear.
+		focus = api.getFocusObject()
+		if len(self.controllerFor)>0:
```
I think you'd want to check that self is focus before checking len(self.controllerFor)... You already fetched focus but didn't use it.

josephsl replied:
When I tested it with Settings and what not, just using self.controllerFor was enough, but I'd be happy to check the focus to really make sure (thanks for catching this).

josephsl committed:
Search Field: check if search field is focused before asking for controller for object. re #6241

Review from Mick Curran (NV Access): Do not waste a function call for fetching focused element. Besides, make sure to check if the focused control is the search field before proceeding to announce appearance of suggestions.

michaelDCurran commented:
@josephsl: how did you go with removing the call to self.parent when detecting suggestionLists?  If it is necessary still, then you should use raw UIA calls to stop possible NVDAObject recursion which could be a  large performance hit. 
Also, it would be good to abstract out the announcements of the opening and closing of suggestions into a behavior, perhaps with some kind of method or event each implementation can call.

josephsl committed:
Update copyright years, add a docstring to UIA objects (__init__).

josephsl committed:
Suggestion: a new behavior to allow NVDA to detect and announce suggestions. re #6241.

Reviewed by Mick Curran (NV Access): announcement of search suggestion is something that NVDA should handle for other API's, such as suggestions in Firefox address bar, search suggestions in universal apps and so on.
A new behavior named 'Suggestion' has been added that allows subclasses to provide custom routines when suggestions appear and disappear. This is handled by event_suggestionOpened and event_suggestionClosed, and by default NVDA will speak and braille this event.

josephsl commented:
Hi,

Behaviors: I agree - after thinking about it, suggestions detection display is something that other API's would like to participate in (not only Start search box in Windows 10, but also address bar in Firefox, Tell Me suggestions in Office 2016 and so on). Search Field in UIA (Windows 10) will now use this behavior.

Thanks.

josephsl committed:
UIA objects/Search field: use the new NVDAObjects.behaviors.Suggestion to announce suggestion appearance. re #6241.

Reviewed by Mick Curran (NV Access): Based on the new Suggestion behavior mix-in, it is now possible for various objects to provide custom routines to let users know the appearance of suggestions. Thus UIA/Search Field is the first object to use this routine.

josephsl commented:
Hi,

Regarding using self.parent: yes, this is the case. I'll try your suggestion of using UIA calls directly. Thanks.

derekriemer commented:
Note:
UIAHandler.handler.baseTreeWalker is your friend, if you can't use
getCachedParent

On Tue, Jan 17, 2017 at 10:39 PM, Joseph Lee <notifications@github.com>
wrote:

> Hi,
>
> Regarding using self.parent: yes, this is the case. I'll try your
> suggestion of using UIA calls directly. Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/6274#issuecomment-273388210>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFGivd8nLFMSyI5plUb7QAENq_FYSOCoks5rTaWWgaJpZM4JmDqm>
> .
>



-- 

Derek Riemer: Improving the world one byte at a time!

   - University of Colorado Boulder Department of computer science, 4th
   year undergraduate student.
   - Accessibility enthusiast.
   - Proud user of the NVDA screen reader.
   - Open source enthusiast.
   - Skier.

Personal website <http://derekriemer.com>

josephsl committed:
UIA objects/suggestions list: use raw UIA base tree walker to obtain suggestion list (parent). re #6241.

Reviewed by Mick Curran (NV Access): it is better to use raw UIA for obtaining parent element. However, one must be careful to catch COM and value errors (COM because the element might not be there, and value because NULL pointer access is logged). The raw UIA method was also recommended by Derek Riemer.

josephsl committed:
UIA objects: updated comments, removed reference to Redstone.

josephsl committed:
UIA objects/Search field: don't allow suggestions detection be announced in Windows 10 Start menu for consistency with earlier versions of Windows. re #6241.

iN Windows 10 Start menu, suggestions are announced automaticlaly, so no need to provide suggestion announcements.

josephsl committed:
NVDAObjects/behaviors: rename 'Suggestion' to 'EditableTextWithSuggestions' so the behavior can be better described. re #6241.

If one inherits from 'Suggestion', the overall impression would be that the field is only going to display suggestions nad nothing else. Many suggestions are shown when text is entered, thus it is better to say 'EditableTextWithSuggestions' to better reflect what the behavior actually does.
Also, a sound will be played to let users know the appearance of suggestions. (default vlaue).

josephsl committed:
Auto-suggestions: add a combo box in object presentation settings to configure how auto-suggestions should be announced. re #6241

josephsl committed:
User guide: document auto-suggestions notification setting. re #6241.

In the user guide, an explanatory text has been added to describe what auto-suggestions are.

josephsl commented:
Attached is a zip file used for announcing appearance/disappearance of auto-suggestions (credit: @DerekRiemer). Thanks.
[nvda_i6241waves.zip](https://github.com/nvaccess/nvda/files/715717/nvda_i6241waves.zip)

michaelDCurran commented:
A few thoughts:
1. @jcsteh has suggested that if we have a sound, then it should only ever be a sound, and not to allow configuring for speech. Thus removes the need for the combo box in settings. I think I agree.
2. I don't think it is necessary to deliberately stop it from  occurring in the Start Screen. In fact if it is only a sound it does not change existing speech, and helps the user to be come more familiar with the sound.
3. What about Microsoft Edge Address bar?  I think its automationID is addressEditBox. However, I'm not sure  what their suggestions list is or if it will match. But we'd definitely want to support it.

josephsl committed:
Auto-suggestions: either play a sound or do nothing. re #6241.

Comments from Mick Curran and Jamie Teh (NV Access): no need for a separate message option when announcing appearance of auto-suggestions, as the sound cue will let the user of this fact. However, braille users should be notified of this regardless of this flag being turned on (deaf-blind users should be notified of this).
User guide and settings dialog: changed the control type and label for auto-suggestions setting to reflect change in behavior.

josephsl committed:
Search suggestions: allow Windows 10 Start menu and Edge to provide search suggestions. re #6241.

Reviewed by Mick Curran (NV Access): allow Start menu search box and Edge's address omnibar to participate in providing auto-suggestions announcement.

josephsl committed:
Config: auto-suggestion setting is now part of configspec module. re #6241

jcsteh commented:
As noted in https://github.com/nvaccess/nvda/issues/6753#issuecomment-273935935, while I'm grateful to @derekriemer for offering his sounds, I think we need to find sounds that are more subtle and less likely to distract the user or drown out simultaneous speech.

michaelDCurran committed:
Merge branch 'master' into pr/6274

michaelDCurran committed:
Update to miscDeps containing auto suggestions sounds.

josephsl committed:
Merge branch 'master' into i6241UIASuggestionListItem

josephsl committed:
NVDAObjects.UIA: catch UnboundLocalError for parentElement if parentElement fetcher fails. re #6241.

In some cases, when Start menu opens, it isn't announced by NVDA. As a result, parent element fetcher fails when trying to instantiate suggestions list item, with a traceback that ends with UnboundLocalError. Catch this by moving the SuggestionsListItem selector to inside of the try block.

josephsl committed:
Merge branch 'master' into i6241UIASuggestionListItem

josephsl committed:
EditableTextWithSuggestions: remove braille message shown when suggestions close for consistency with other situations (such as browse mode toggle). re #6241.

josephsl committed:
NVDAObjects.UIA.SuggestionListItem: braille suggestion results as flash messages. re #6241.

Suggestion from Davy Kager: provide a way to let braille users read search suggestion items. This is done by emulating some parts of speech.SpeakObjectProperties except the name and position info map will be fetched (position info map fetching is contingent on whether report position info setting is enabled from Object Presentation dialog). Ideally, NVDA objects should have a way to fetch braille flash messages for controls.
Also reworded docstring for SuggestionListItem so it cna include other UIA-based suggestion list items such as Windows 8.x search results.

josephsl committed:
NVDAObjects.UIA.SuggestionsListItem: flash suggestion results in braille. re #6241, #6414.

Instead of constructing the likely flash message, use a function used as part of Core issue 6414, which is much simpler than constructing the flash message from scratch.

jcsteh commented:
I could swear this was working the other day, but maybe I was running the wrong build. STR:

1. Open the Start menu.
2. Search for an app.
3. Press the applications key.
4. Use the up and down arrows to move through the context menu items.
    * Expected: NVDA should report items like Pin to Start, Pin to taskbar, etc.
    * Actual: It repeatedly reports what you typed into the search box and nothing more.

I think the UIA suggestion list item code needs to be extended to cover items with a parent automation id of "contextMenu".

josephsl commented:
Hi, it reports context menu items along with what you’ve typed into the search field, I believe same behavior as in 2017.2. Thanks.

 

From: James Teh [mailto:notifications@github.com] 
Sent: Wednesday, June 14, 2017 4:58 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] UIA objects: introduce a global suggestion list item class for Windows 10. re #6241 (#6274)

 

I could swear this was working the other day, but maybe I was running the wrong build. STR:

1.	Open the Start menu.
2.	Search for an app.
3.	Press the applications key.
4.	Use the up and down arrows to move through the context menu items. 

*	Expected: NVDA should report items like Pin to Start, Pin to taskbar, etc.
*	Actual: It repeatedly reports what you typed into the search box and nothing more.

I think the UIA suggestion list item code needs to be extended to cover items with a parent automation id of "contextMenu".

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6274#issuecomment-308409277> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkCOpnxu2zT5xEBl056mStLCd0M69ks5sD8rdgaJpZM4JmDqm> .

jcsteh commented:
Mick just confirmed this doesn't work for him either. I just figured out
that it works if Windows 10 App Essentials is enabled. If it's disabled, it
doesn't work. However, it does work in master. git bisect also shows that
this PR causes the regression. From the code, I can also see why this
wouldn't work: the suggestion item class only gets used for a specific
parent automation id.

josephsl commented:
Oh great… Just confirmed. An emergency patch is on its way. Thanks.

 

From: James Teh [mailto:notifications@github.com] 
Sent: Wednesday, June 14, 2017 5:42 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] UIA objects: introduce a global suggestion list item class for Windows 10. re #6241 (#6274)

 

Mick just confirmed this doesn't work for him either. I just figured out
that it works if Windows 10 App Essentials is enabled. If it's disabled, it
doesn't work. However, it does work in master. git bisect also shows that
this PR causes the regression. From the code, I can also see why this
wouldn't work: the suggestion item class only gets used for a specific
parent automation id.



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6274#issuecomment-308596584> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkMnrmiE9qWltJ-t1iyaXH-azrGtYks5sEH3MgaJpZM4JmDqm> .

josephsl committed:
Merge branch 'master' into i6241UIASuggestionListItem

josephsl committed:
UIA/SuggestionListItem/searchui: also classify Start search results context menu as a suggestions list item. re #6241.

Oddly, the same behavior that's applied to suggestion items must work in Start suggestion's context menu, otherwise menu items will not be announced.

michaelDCurran commented:
@josephsl: could you please merge master and resolve the conflict. It is just the addition of the docstring to ```NVDAObjects.UIA```. Then I can merge this to master. Thanks.

josephsl committed:
Merge branch 'master' into i6241UIASuggestionListItem

josephsl commented:
Hi, done. I added docstring there as part of improving the source code documentation. Thanks.

 

From: Michael Curran [mailto:notifications@github.com] 
Sent: Thursday, June 22, 2017 10:35 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] UIA objects: introduce a global suggestion list item class for Windows 10. re #6241 (#6274)

 

@josephsl <https://github.com/josephsl> : could you please merge master and resolve the conflict. It is just the addition of the docstring to NVDAObjects.UIA. Then I can merge this to master. Thanks.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6274#issuecomment-310576054> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkHG9HR4R3vSLmQE4ryGDwky74iP2ks5sG06JgaJpZM4JmDqm> .
